src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -110,12 +118,62 @@ protected Object readResolve() throws ObjectStreamException {
      * It will also be false for something that has finished but is pending child node creation,
      * such as a completed fork branch which is waiting for the join node to be created.
      * <p>This can only go from true to false and is a shortcut for {@link FlowExecution#isCurrentHead}.
+     * @see #isActive
      */
     @Exported
     public final boolean isRunning() {
         return getExecution().isCurrentHead(this);
     }
 
+    private static final Map<FlowExecution, Set<BlockStartNode>> unclosedBlocks = new WeakHashMap<>();
+    @Restricted(DoNotUse.class)
+    @Extension public static class BlockListener implements GraphListener.Synchronous {
+        @Override public void onNewHead(FlowNode node) {
+            if (node instanceof BlockStartNode || node instanceof BlockEndNode) {
+                FlowExecution exec = node.getExecution();
+                synchronized (unclosedBlocks) {
+                    Set<BlockStartNode> blocks = unclosedBlocks.get(exec);
+                    if (blocks == null) {
+                        unclosedBlocks.put(exec, blocks = new HashSet<>());
+                    }
+                    if (node instanceof BlockStartNode) {
+                        blocks.add((BlockStartNode) node);
+                    } else {
+                        blocks.remove(((BlockEndNode) node).getStartNode());
+                    }
+                }
+            }
+        }
+    }
+    /**
+     * Checks whether a node is still part of the active part of the graph.
+     * Unlike {@link #isRunning}, this behaves intuitively for a {@link BlockStartNode}:
+     * it will be considered active until the {@link BlockEndNode} is added.
+     */
+    public final boolean isActive() {
+        if (this instanceof FlowEndNode) { // cf. JENKINS-26139
+            return false;
+        } else if (this instanceof BlockStartNode) { // cf. JENKINS-38223
+            synchronized (unclosedBlocks) {
+                Set<BlockStartNode> blocks = unclosedBlocks.get(exec);
+                if (blocks != null) {
+                    return blocks.contains((BlockStartNode) this);
+                } else {
+                    // Need workflow-cps 2.33+ to use the optimization.
+                    LOGGER.log(Level.FINE, "falling back to old isActive implementation for {0}", this);
+                    if (isRunning()) {
+                        return true;
+                    }
+                    List<FlowNode> headNodes = exec.getCurrentHeads();
+                    AbstractFlowScanner scanner = (headNodes.size() > 1) ? new DepthFirstScanner() : new LinearBlockHoppingScanner();
+                    return scanner.findFirstMatch(headNodes, Predicates.equalTo(this)) != null;
+                }
```

Comment:
Future enhancement if we do something like this: if we gotta scan the whole flow, fully populate the cache of the block structure along the way, so we never have to do that full scan again.